pkgs/build-support/docker/default.nix

```diff
@@ -75,6 +76,13 @@ let
   # mapping from the go package.
   defaultArch = go.GOARCH;
 
+  # Add ARM variant support (e.g. v8, v7...)
+  defaultVariant =
+    if stdenv.hostPlatform.parsed.cpu.version != null then
```

Comment:
```suggestion
  defaultVariant = arch:
    if stdenv.hostPlatform.parsed.cpu.version != null && arch == defaultArch then
```

Reply:
`defaultVariant` function is used in different different scopes and some are making no sense to use the check `arch == defaultArch` since the value of the arch in these function is already defined by defaultArch (see the links below).

https://github.com/DarkOnion0/nixpkgs/blob/f818cb157252a2c7aa2faac5847b189a901501ee/pkgs/build-support/docker/default.nix#L553

https://github.com/DarkOnion0/nixpkgs/blob/f818cb157252a2c7aa2faac5847b189a901501ee/pkgs/build-support/docker/default.nix#L862

What about only adding this check in the `pullImage` function (the only function where the user can set its own arch)  with an error message if `arch == defaultArch` is not true ?

Reply:
> What about only adding this check in the pullImage function (the only function where the user can set its own arch)

Sounds good.

> with an error message if arch == defaultArch is not true

I don't see why this would be necessary. How would you fetch the right arm image when hostPlatform is x86?
This is a little weird, but changing `arch` is the purpose of the `arch` flag.

What I think we really want is to treat the combination of `arch` and `variant` as one parameter, so that it is either completely based on `hostPlatform`, or completely based on direct user input.

We should also consider deprecating these parameters in favor of `hostPlatform`, which is how the other functions work. Looking at its long git history, it seems that this logic was added in the form of parameters as a quick and dirty shortcut to make things work, rather than as an intentional feature.
The same control will remain possible via the normal cross compilation support.

Reply:
> stdenv.hostPlatform.parsed.cpu.version 

This seems to duplicate the `go.GOARM` logic, which also shows that `version` may be unset.
It seems better to derive `defaultVariant` from `GOARM`.

Reply:
> This seems to duplicate the `go.GOARM` logic, which also shows that `version` may be unset.

You're right, I did this originally because the `GOARM` var doesn't support the `v8` arm version but after further reading, the `v8` variant is optional, knowing that I'll be migrating the `defaultVariant` to `go.GOARM`.

Reply:
> I don't see why this would be necessary. How do you get the right arm image when hostPlatform is x86?
> It's a bit weird, but changing `arch` is the purpose of the `arch` flag.

I make this proposition to prevent the user from compiling a container for another operating system than the nixpkgs host version if they don't use the cross-compiling options built into nix (e.g. `pkgs.pkgsCross.aarch64-multiplatform`...).

But, as you proposed:

> We should also consider deprecating these parameters in favor of `hostPlatform`, which is how the other functions work. Looking at its long git history, it seems that this logic was added in the form of parameters as a quick and dirty shortcut to make things work, rather than as an intentional feature.
> The same control will remain possible via the normal cross compilation support.

I will work on deprecating these flags and implement the same logic as in the other functions